### ISSUE TITLE
Including both shadow copy and  deep copy  may be better

### ISSUE THREAD

**Issue by wangtengchao** [mapstruct/mapstruct#695]
Sometimes,developer need shadow copy  and somtimes deep copy may be a good choice .  If mapstruct support both of them, and developer can configure it ,i think it's better. I don't know whether this fuction has already been developed and haven't found it in document. 
Thank you!

**Comment by gunnarmorling** [comment]
Hey @wangtengchao, could you describe a bit more what exactly you mean by "shadow copy" and "deep copy", just to make sure I get what you are after. Maybe you could add an example, specifically of the code you would like to see generated? Thanks!

**Comment by wangtengchao** [comment]
class Product{
     private Attribute attribute;
}
class ProductMapper{
    ProductDTO productToProductDTO(Product product)
}
=>
1. Now the code is generated like this:
class ProductMapperImpl implments ProductMapper{
    ProductDTO productToProductDTO(Product product){
         ProductDTO productDTO = new ProductDTO();
         productDTO.setAttribute(product.getAttribute()); // shadow copy
   }
}
2. But sometimes, we need things below  
Attribute attribute = new Attribute(); // deep copy . generate a new object
attribute.setId(product.getAttribute().getId()); //copy data
 productDTO.setAttribute(attribute); // set the new object to productDTO

**Comment by gunnarmorling** [comment]
Could you try and write a method yourself which does the cloning and add it via a mapper imported with `@Mapper#users()`:

    public Attribute clone(Attribute attribute) {
        return new Attribute(attribute.getId());
    }

The generated mapper then should pick it up when converting the `attribute` property in the `productToProductDTO()` method.

**Comment by wangtengchao** [comment]
Yes, It's a good way  to realize it. But I think this is a common function for data mapping  , it may be better if i can configure it by annotation or else. 
Thanks!

**Comment by ebelanger** [comment]
I would add my approval for such a change.  It would be very nice indeed if MapStruct could be used to create cloning methods since it would make sure that no properties are missed VS writing the method manually.

**Comment by ebelanger** [comment]
I think I have found a way to use MapStruct to generate such cloning methods, but it will require 2 methods.

``
public Attribute clone(Attribute attribute) {
    return clone(source, new Attribute());
}

protected abstract Attribute clone(Attribute source, @MappingTarget Attribute target);
``

**Comment by amidudu** [comment]
Not sure if something changed 2 years after...
We are using MapStruct for cloning and faced the same issue.
In our cases we clone an object which has list of a different class, the list in the cloned object is a new list but the objects inside are not cloned. 
We created a clone for the inner class but the generation doesn't use it, tried both solutions described above without success.

Here is an example:
```java
class A {
	private List<B> list;
...
}

class B {
...
}

@Mapper(componentModel = "spring")
public interface ACloner {

	A clone(A a);
	B clone(B b);
}
```
The output looks something like this (doesn't use the B clone method)
```java
@Generated(
    value = "org.mapstruct.ap.MappingProcessor"
)
@Component
public class AClonerImpl implements ACloner {

    @Override
    public A clone(A a) {
        if ( a == null ) {
            return null;
        }

        A a1 = new A();

        List<B> list1 = a.getList();
        if ( list1 != null ) {
            a1.setList( new ArrayList<B>( list1 ) );
        }
        else {
            a1.setList( null );
        }
...
}
```

Any solution/workaround?

**Comment by sjaakd** [comment]
@amidudu MapStruct has different behaviour if the elements  are the same (example of your above) or different. When arguments are different, MapStruct creates its own iterable mapping method.  MapStruct will also call such method if you provide the signature.

I was just thinking: what if you provide an additional iterable mapping method signature. Something like this:

```java
@Mapper(componentModel = "spring")
public interface ACloner {

	A clone(A a);
	List<B> clone(List<B> b);
}
```

I'm not sure whether you (even) have to provide  a mapping `B clone( B (b)`.. Have a shot at it and let me know.

**Comment by amidudu** [comment]
@sjaakd Thanks, this is a good idea and works if you have one list, we have multiple list so the methods has same erasure, so this will require to split the mapper interface into multiple ones, which will make it less readable.

Also there is an general flaw with this (and my) implementation - you need to remember to update the interface when the class changes and additional list/class is added to A. 
I think that adding annotations for "deep clone" can be the best solution.

What do you think?

**Comment by sjaakd** [comment]
> What do you think?

I agree. That's why I kept the issue open :smile:. The solution should not be that complex. Although we need to step back a little and re-consider `Mapper#disableSubMappingMethodsGeneration `.. In the end I think its a sliding scale with at one end of the spectrum: 
1. `disableSubMappingMethodsGeneration = true` (no automatic nesting sub-mapping generation on bean mappings, 
2. `disableSubMappingMethodsGeneration = false` (default). Sub-mapping generations on bean mappings 
3. `disableSubMappingMethodsGeneration = false ++`. Force sub-mapping everywhere (bean-mapping-properties when properties are the same, map-mapping, iterable-mapping).

So possibly this `disableSubMappingMethodsGeneration` has to be replaced with a downward incompatible enum (`SubMappingStrategy`).  

@filiphr WDYT?
@amidudu .. if you want to help out.?

**Comment by filiphr** [comment]
Adding a cloning option is certain something interesting.  Your proposal with a `SubMappingStrategy` could be a way of doing it. I don't see a big problem in our side about adding this

**Comment by johnzielke** [comment]
Is there any progress on this?

**Comment by sjaakd** [comment]
I was thinking: in `MapperResolverImpl` we check which method of mapping to apply:
1. Direct
2. Method
3. Conversion
4. Method/Conversion, Method/Method

What if we made a strategy (some kind of mask) in order to determine which steps to apply.  Note: for deep cloning we only apply 2. There has been an earlier request to switch of 3 & 4.

The idea could be to use an enum which has a second value to mask (byte) . Default is ALL.

**Comment by chenfangzhi123** [comment]
any progress ?

**Comment by sjaakd** [comment]
> any progress ?

still pending on available spare time :smile:

**Comment by chenfangzhi123** [comment]
> > any progress ?
> 
> still pending on available spare time 😄

looking forward to😄

**Comment by sjaakd** [comment]
@wangtengchao .. it took a long time.. but it's there.

@chenfangzhi123 give it a spin.. Its on our master.Let is know.  I'll proceed to close this issue.

**Comment by filiphr** [comment]
For documentation purposes. In order to do a deep clone you would need to use the new `MappingControl` on the `Mapping`, `BeanMapping`, `IterableMapping`, `MapMapping`, `Mapper` or `MapperConfig`.

For deep cloning it would be:

```
import org.mapstruct.control.DeepClone;

interface ProductMapper {

    @Mapping(target = "attribute", mappingControl = DeepClone.class)
    ProductDTO productToProductDTO(Product product)
}
```

If you want to clone the whole object then:


```
import org.mapstruct.control.DeepClone;

@Mapper(mappingControl = DeepClone.class)
interface ProductMapper {

    ProductDTO cloneProduct(ProductDto product)
}
```

**Comment by patrickrioux** [comment]
Hi,

Where is that org.mapstruct.control.DeepClone; class from the previous post?

If the source is a Hibernate Entity object being used to generated a DTO, this should be a clone copy of the DTO. But it sounds like that DTO is a copy of the Entity which is still attached to the DB. 

I am not quite sure if mapstruct is the right tool.

**Comment by filiphr** [comment]
@patrickrioux this feature would be part of the 1.4 release. Currently you can try it by using 1.4.0-SNAPSHOT.

>If the source is a Hibernate Entity object being used to generated a DTO, this should be a clone copy of the DTO. But it sounds like that DTO is a copy of the Entity which is still attached to the DB.

Sorry but I couldn't understand what you mean with that. Can you give an example perhaps?

**Comment by govindrk** [comment]
Hey @filiphr,

Sorry about the git issue that i created.

I have a very complex nested (very deep) Hibernate JPA entity with many relationship (one-to-one, one-to-many and many-to-one). I want to copy only the Id from one entity to another entity of same complex structure.

SubjectEntity  subjectEntity
SubjectEntity updatedEntity

The first subjectEntity will hold the primary key (Id) from the database table. The updatedEntity which is an update to an existing record is created from JAXB to SubjectEntity without the Id element.

When i save the updatedEntity directly without Id (null), the existing record is deleted and new record is created. I have to query for an existing record and copy only the Id into updatedEntity. There are around 280 Ids.

Please let me know if this can be achieved through mapstruct.
Thanks

**Comment by filiphr** [comment]
Hey @govindrk,

My reply to the issue was not only because you created an issue. Our issue tracker is for bugs, enhancements and new features for MapStruct. Not for answering questions. Please post your question with examples of what you've tried on StackOverflow or our [Gitter Room](https://gitter.im/mapstruct/mapstruct-users).

**Comment by govindrk** [comment]
> Hey @govindrk,
> 
> My reply to the issue was not only because you created an issue. Our issue tracker is for bugs, enhancements and new features for MapStruct. Not for answering questions. Please post your question with examples of what you've tried on StackOverflow or our [Gitter Room](https://gitter.im/mapstruct/mapstruct-users).

My apologies.

**Comment by z0mb1ek** [comment]
when will it be released?

**Comment by nbrugger-tgm** [comment]
Hi, while I enjoy that there was work done and the `@DeepClone` annotation was introduced, there is not much of an explanation of what it actually does. Maybe I am just bad at finding it, but neither the source code nor the Javadoc reveals what adding `@Mapper(controll = DeepClone.class)` changes

**Comment by filiphr** [comment]
@nbrugger-tgm if something is not clear can you please open a new issue and let us know what and where we can write something to make it clear?

Currently `DeepClone` has the following javadoc:

>Clones a source type to a target type (assuming source and target are of the same type).

There is also extensive Javadoc on `MappingControl`
